src/pretix/base/datasync/datasync.py

```diff
@@ -0,0 +1,425 @@
+        }
+
+    def get_field_value(self, inputs, mapping_entry):
+        key = mapping_entry["pretix_field"]
+        try:
+            required_input, label, ptype, enum_opts, getter = self.data_fields[key]
+        except KeyError:
+            raise SyncConfigError(['Field "%s" is not valid for %s. Please check your %s settings.' % (key, "/".join(inputs.keys()), self.display_name)])
+        input = inputs[required_input]
+        val = getter(input)
+        if isinstance(val, list):
+            if enum_opts and mapping_entry.get("value_map"):
+                map = json.loads(mapping_entry["value_map"])
+                try:
+                    val = [map[el] for el in val]
+                except KeyError:
+                    raise SyncConfigError([f'Please update value mapping for field "{key}" - option "{val}" not assigned'])
+
+            val = ",".join(val)
+        return val
+
+    def get_properties(self, inputs: dict, property_mapping: str):
+        property_mapping = json.loads(property_mapping)
+        return [
+            (m["external_field"], self.get_field_value(inputs, m), m["overwrite"])
+            for m in property_mapping
+        ]
+
+    """
+    This method is called for each object that needs to be created/updated in the external system -- which these are is
+    determined by the implementation of the `mapping` property.
+
+    :param pk_field: Identifier field in the target system as provided in mapping.external_pk
+    :param pk_value: Identifier contents as retrieved from the property specified by mapping.pretix_pk of the model
+                     specified by mapping.pretix_model
+    :param properties: All properties defined in mapping.property_mapping, as list of three-tuples
+                       (external_field, value, overwrite)
+    :param inputs: All pretix model instances from which data can be retrieved for this mapping
+    :param mapping: The mapping object as returned by self.mappings
+    :param mapped_objects: Information about objects that were synced in the same sync run, by mapping definitions
+                           *before* the current one in order of self.mappings.
+                           Type is a dictionary {mapping.pk: [list of return values of this method]}
+                           Useful to create associations between objects in the target system.
+    :return: {
+                "object_type": mapping.external_object_type,
+                "pk_field": pk_field,
+                "pk_value": pk_value,
+                "external_link_href": "https://external-system.example.com/backend/link/to/contact/123/",
+                "external_link_display_name": "Contact #123 - Jane Doe",
+                "...optionally further values you need in mapped_objects for association": 123456789,
+             }
+
+    This method needs to be idempotent, i.e. calling it multiple times with the same input values should create
+    only a single object in the target system.
```

Comment:
I don't think this will be possible for all target systems. For newsletter systems and CRMs yes, but for accounting systems … I'm cautious it might not be possible :\ How hard do we depend on this if we have the locking and e.g. a plugin listens only on order_placed?

Reply:
-> https://github.com/pretix/pretix/pull/4881#issuecomment-2855118900